### PR TITLE
multus: add multus manifests

### PR DESCRIPTION
**Pull request by zshi-redhat**
- add manifests for multus-cni, sriov-cni, sriov-device-plugin
- add Namespace and Annotations fields in AdditionalNetworkDefinition structure to enable creation of namespace isolated and annotated Custom Resource (annotated CR is used to pass SR-IOV resource annotation to multus).
- add multusNodeConfig function to generate content of Multus CNI configMap: 70-multus.conf

### CODE REVIEW

**Comment by zshi-redhat** [comment]
/assign @squeed  Hi Casey, this is a follow up PR to add multus related manifests, please help to review, Thanks!

@s1061123 @dougbtv please help to review if multus manifests are correct, I copied them from downstream repo and did a little modification to fit with openshift.
This is to add manifests for multus-cni, sriov-cni, sriov-device-plugin. I think we can add more (supported/unsupported containernteworking cni, admission controllers ) if related manifests are ready to consume.

With current PR, I occasionally saw a problem of applying CR of net-attach-def Kind:
```
could not apply (k8s.cni.cncf.io/v1, Kind=NetworkAttachmentDefinition) multus/sriov-net: could not retrieve existing (k8s.cni.cncf.io/v1, Kind=NetworkAttachmentDefinition) multus/sriov-net: no matches for kind "NetworkAttachmentDefinition" in version "k8s.cni.cncf.io/v1"
```
And this is the yaml file I used to created NetworkConfig custom resource:
```
apiVersion: "networkoperator.openshift.io/v1"
kind: "NetworkConfig"
metadata:
  name: "default"
spec:
  serviceNetwork: "172.30.0.0/16"
  clusterNetworks:
    - cidr: "10.128.0.0/14"
      hostSubnetLength: 9
  defaultNetwork:
    type: OpenShiftSDN
    openshiftSDNConfig:
      mode: NetworkPolicy
  additionalNetworks:
    - type: Raw
      name: sriov-net
      namespace: multus
      annotations:
        k8s.v1.cni.cncf.io/resourceName: intel.com/sriov
      rawCNIConfig: '{
        "type": "sriov",
        "name": "sriov-network",
        "namespace": multus,
        "ipam": {
                "type": "host-local",
                "subnet": "10.56.217.0/24",
                "routes": [{
                        "dst": "0.0.0.0/0"
                }],
                "gateway": "10.56.217.1"
        }
}'
```
do you see what might be the problem?

**Comment by zshi-redhat** [comment]
updated code gen and verified code style.

**Comment by zshi-redhat** [comment]
the CI test failure doesn't seem to be related with network operator,  re-run the tests:
/retest

File: bindata/network/additional-networks/multus/multus.yaml [hunk 1]
```diff
@@ -0,0 +1,72 @@
+kind: ConfigMap
```
[the remark this answers is not in the document]
Comment (zshi-redhat) [reply bindata/network/additional-networks/multus/multus.yaml]: Good catch! Thanks Tomo for the review!
fixed and squashed the commit.

**Comment by zshi-redhat** [comment]
> /assign @squeed Hi Casey, this is a follow up PR to add multus related manifests, please help to review, Thanks!
> 
> @s1061123 @dougbtv please help to review if multus manifests are correct, I copied them from downstream repo and did a little modification to fit with openshift.
> This is to add manifests for multus-cni, sriov-cni, sriov-device-plugin. I think we can add more (supported/unsupported containernteworking cni, admission controllers ) if related manifests are ready to consume.
> 
> With current PR, I occasionally saw a problem when the operator applys CR of net-attach-def Kind:
> 
> ```
> could not apply (k8s.cni.cncf.io/v1, Kind=NetworkAttachmentDefinition) multus/sriov-net: could not retrieve existing (k8s.cni.cncf.io/v1, Kind=NetworkAttachmentDefinition) multus/sriov-net: no matches for kind "NetworkAttachmentDefinition" in version "k8s.cni.cncf.io/v1"
> ```
> And this is the yaml file I used to created NetworkConfig custom resource:
> 
> ```
> apiVersion: "networkoperator.openshift.io/v1"
> kind: "NetworkConfig"
> metadata:
>   name: "default"
> spec:
>   serviceNetwork: "172.30.0.0/16"
>   clusterNetworks:
>     - cidr: "10.128.0.0/14"
>       hostSubnetLength: 9
>   defaultNetwork:
>     type: OpenShiftSDN
>     openshiftSDNConfig:
>       mode: NetworkPolicy
>   additionalNetworks:
>     - type: Raw
>       name: sriov-net
>       namespace: multus
>       annotations:
>         k8s.v1.cni.cncf.io/resourceName: intel.com/sriov
>       rawCNIConfig: '{
>         "type": "sriov",
>         "name": "sriov-network",
>         "namespace": multus,
>         "ipam": {
>                 "type": "host-local",
>                 "subnet": "10.56.217.0/24",
>                 "routes": [{
>                         "dst": "0.0.0.0/0"
>                 }],
>                 "gateway": "10.56.217.1"
>         }
> }'
> ```
> do you see what might be the problem?

It looks like that CRD and associated CR cannot be created in a single run, if I restart the operator, the net-attach-def CR can be successfully created (since the net-attach-def CRD is already created before restarting operator). It might imply that we need to apply CRD and CR separately in different context. It doesn't seem that operator supports it now.

**Comment by dcbw** [comment]
@zshi-redhat so the branch I'm working on for Multus support takes a lighter weight approach, see here: https://github.com/openshift/cluster-network-operator/pull/54 . I think that we should keep the Multus stuff separate and not logically combine it with AdditionalNetworks or the SRIOV plugin.

SRIOV can be a separate PR that does actually inspect AdditionalNetworks for SRIOV-enabled ones and then renders the DevicePlugin and associated RBAC/Namespace. But that's logically separate from Multus configuration, and the Multus config can be very simple (eg we don't need to specify Delegates since everything except the default network should be handled by AdditionalNetworks).

Thoughts?

**Comment by zshi-redhat** [comment]
> @squeed @zshi-redhat so the branch I'm working on for Multus support takes a lighter weight approach, see here: #54 . I think that we should keep the Multus stuff separate and not logically combine it with AdditionalNetworks or the SRIOV plugin.

+1 to keep Multus config separate from SR-IOV and additionalNetworks.
This will also be useful when later we want to enable migration between different network solutions.

> 
> SRIOV can be a separate PR that does actually inspect AdditionalNetworks for SRIOV-enabled ones and then renders the DevicePlugin and associated RBAC/Namespace. But that's logically separate from Multus configuration, and the Multus config can be very simple (eg we don't need to specify Delegates since everything except the default network should be handled by AdditionalNetworks).

I remember Multus now supports config default network with net-attach-def CR, in which case it may re-use the additional Network CR template to render the default network during Multus configuration, this may requires additional network be configured to provide default network CR.

> 
> Thoughts?

**Comment by dcbw** [comment]
Should we close this one and do separate follow-ups for any parts that aren't already there?

**Comment by zshi-redhat** [comment]
> Should we close this one and do separate follow-ups for any parts that aren't already there?

Yes, this should be closed. Thanks for the reminding!

**Pull request closed without merging**